File: demos/python/basic_ops.py
"""Basic operations demo for Insight7 Python bindings.

Demonstrates: array creation, arithmetic, multi-dtype, broadcasting,
unary operations, and reductions.
"""

import sys
import os
import math

# Add bindings to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "bindings", "python"))
sys.path.insert(
    0, os.path.join(os.path.dirname(__file__), "..", "..", "build", "bindings", "python")
)

import insight as ins
import numpy as np


def separator(title):
    print(f"\n{'=' * 40}")
    print(f"  {title}")
    print(f"{'=' * 40}")


def main():
    ins.init()

    print("Insight7 Basic Operations Demo (Python)")

    # --- Array creation ---
    separator("Array Creation")

    a = ins.ones(ins.Shape([3, 4]), ins.float32)
    s = a.shape()
    print(f"ones(3,4) shape=({s[0]},{s[1]}) dtype={a.dtype()}")

    b = ins.arange(12, dtype=ins.float64).reshape(ins.Shape([3, 4]))
    print(f"arange(0,12).reshape(3,4):\n{b}")

    c = ins.eye(4, dtype=ins.float64)
    print(f"eye(4) F64:\n{c}")

    # --- Multi-dtype arithmetic ---
    separator("Multi-Dtype Arithmetic")

    # F32
    f32_a = ins.from_array([1.0, 2.0, 3.0]).to(ins.float32)
    f32_b = ins.from_array([4.0, 5.0, 6.0]).to(ins.float32)
    f32_c = f32_a + f32_b
    print(f"F32: [1,2,3] + [4,5,6] = {f32_c}")

    # F64
    f64_a = ins.from_array([1.0, 2.0, 3.0])
    f64_b = ins.from_array([0.5, 0.5, 0.5])
    f64_c = f64_a * f64_b
    print(f"F64: [1,2,3] * [0.5,0.5,0.5] = {f64_c}")

    # I32
    i32_a = ins.from_array([10, 20, 30]).to(ins.int32)
    i32_b = ins.from_array([3, 3, 3]).to(ins.int32)
    i32_c = i32_a / i32_b
    print(f"I32: [10,20,30] / [3,3,3] = {i32_c.to(ins.int32)}")

    # --- Broadcasting ---
    separator("Broadcasting")

    m = ins.from_array([[1.0, 2.0, 3.0], [4.0, 5.0, 6.0]])
    row = ins.from_array([[10.0, 20.0, 30.0]])
    result = m + row
    print(f"Matrix + Row broadcast:\n{result}")

    # --- Unary operations ---
    separator("Unary Operations")

    x = ins.from_array([-2.0, -1.0, 0.0, 1.0, 2.0])
    print(f"x:      {x}")
    print(f"abs(x): {ins.abs(x)}")
    print(f"sin(x): {ins.sin(x)}")
    print(f"exp(x): {ins.exp(x)}")

    # --- Reductions ---
    separator("Reductions")

    data = ins.from_array([1.0, 2.0, 3.0, 4.0, 5.0])
    print(f"data:  {data}")
    print(f"sum:   {ins.sum(data)}")
    print(f"mean:  {ins.mean(data)}")
    print(f"max:   {ins.max(data)}")
    print(f"min:   {ins.min(data)}")

    print("\nDone!")


if __name__ == "__main__":
    main()

File: demos/julia/basic_ops.jl
# demos/julia/basic_ops.jl
# Demonstrates: array creation, arithmetic, multi-dtype, broadcasting,
# unary operations, and reductions.

# Add bindings to load path
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "bindings", "julia"))
push!(LOAD_PATH, joinpath(@__DIR__, "..", "..", "build", "bindings", "julia"))

using Insight

function separator(title)
    println("\n" * "="^40)
    println("  $title")
    println("="^40)
end

println("Insight7 Basic Operations Demo (Julia)")

# --- Array creation ---
separator("Array Creation")

a = Insight.ones([3, 4], Insight.float32)
println("ones(3,4) shape=($(Insight.shape(a))) dtype=$(Insight.dtype(a))")

b = Insight.arange(0.0, 12.0, 1.0, Insight.float64)
b = Insight.reshape(b, [3, 4])
println("arange(0,12).reshape(3,4):")
println(b)

c = Insight.eye(Int64(4), Int64(4), Insight.float64)
println("eye(4) F64:")
println(c)

# --- Multi-dtype arithmetic ---
separator("Multi-Dtype Arithmetic")

# F32
f32_a = Insight.cast(Insight.from_data([1.0, 2.0, 3.0]), Insight.float32)
f32_b = Insight.cast(Insight.from_data([4.0, 5.0, 6.0]), Insight.float32)
f32_c = f32_a + f32_b
println("F32: [1,2,3] + [4,5,6] = [$(f32_c[1]),$(f32_c[2]),$(f32_c[3])]")

# F64
f64_a = Insight.from_data([1.0, 2.0, 3.0], Insight.float64)
f64_b = Insight.from_data([0.5, 0.5, 0.5], Insight.float64)
f64_c = f64_a * f64_b
println("F64: [1,2,3] * [0.5,0.5,0.5] = [$(f64_c[1]),$(f64_c[2]),$(f64_c[3])]")

# I32
i32_a = Insight.cast(Insight.from_data([10.0, 20.0, 30.0]), Insight.int32)
i32_b = Insight.cast(Insight.from_data([3.0, 3.0, 3.0]), Insight.int32)
i32_c = i32_a / i32_b
println("I32: [10,20,30] / [3,3,3] = [$(i32_c[1]),$(i32_c[2]),$(i32_c[3])]")

# --- Broadcasting ---
separator("Broadcasting")

m = Insight.from_data(reshape([1.0, 2.0, 3.0, 4.0, 5.0, 6.0], 2, 3), Insight.float64)
row = Insight.from_data(reshape([10.0, 20.0, 30.0], 1, 3), Insight.float64)
result = m + row
println("Matrix + Row broadcast:")
println(result)

# --- Unary operations ---
separator("Unary Operations")

x = Insight.from_data([-2.0, -1.0, 0.0, 1.0, 2.0], Insight.float64)
println("x:      $x")
println("abs(x): $(Insight.abs_fn(x))")
println("sin(x): $(Insight.sin_fn(x))")
println("exp(x): $(Insight.exp_fn(x))")

# --- Reductions ---
separator("Reductions")

data = Insight.from_data([1.0, 2.0, 3.0, 4.0, 5.0], Insight.float64)
println("data:  $data")
println("sum:   $(Insight.sum(data))")
println("mean:  $(Insight.mean(data))")

println("\nDone!")
